### ISSUE TITLE
Navigate to file is slow on large repos

### ISSUE THREAD

**Issue by mdeeks** [microsoft/vscode#145370]
Issue Type: <b>Performance Issue</b>

We have a somewhat large monorepo and searching for files (cmd+p) is quite slow. It takes about 5-7s to search for a basic filename and subsequent searches take about the same amount of time. 

When I watch the process tree I see that ripgrep is called every single time I try to find a file. A comment here in another ticket indicates it should operate off of a cache: https://github.com/microsoft/vscode/issues/142293#issuecomment-1031770183

That doesn't seem to be the behavior I see though. I see the slow search on both local and remote-ssh instances of vscode.

The ripgrep command I see:
`/Applications/Visual Studio Code.app/Contents/Resources/app/node_modules.asar.unpacked/@vscode/ripgrep/bin/rg --files --hidden --case-sensitive -g !**/.git -g !**/.svn -g !**/.hg -g !**/CVS -g !**/.DS_Store -g !**/Thumbs.db -g !**/node_modules -g !**/bower_components -g !**/*.code-search --no-ignore-parent --follow --no-config --no-ignore-global`

VS Code version: Code 1.65.2 (Universal) (c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1, 2022-03-10T14:33:49.188Z)
OS version: Darwin arm64 21.3.0
Restricted Mode: No
Remote OS version: Linux x64 5.13.0-1017-aws

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Pro (10 x 24)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 4, 3|
|Memory (System)|32.00GB (0.95GB free)|
|Process Argv|--crash-reporter-id 02e61eab-6204-4b3a-8524-e05f7728a750|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: bento|
|OS|Linux x64 5.13.0-1017-aws|
|CPUs|Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz (8 x 2499)|
|Memory (System)|30.90GB (23.47GB free)|
|VM|0%|
</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
    9	   262	  4359	code main
    4	   131	  4365	   gpu-process
    0	    33	  4366	   utility-network-service
    0	   229	  4370	   window (main.go — bento-server)
    0	   262	  4371	   window (server.go — bento)
   12	   229	  4886	   shared-process
    0	    33	  5039	     ptyHost
    0	     0	  5772	       /bin/zsh -l
    0	     0	 25185	       /bin/zsh -l
    0	    33	  5214	     fileWatcher
    0	    66	  5215	     fileWatcher
    0	    66	 39855	     fileWatcher
    0	    66	 41504	     fileWatcher
    0	     0	 41794	     /bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	    33	 73743	     fileWatcher
    0	   164	  5641	   extensionHost
    0	   328	 53114	     /Users/michaeldeeks/go/bin/gopls -mode=stdio
    0	   131	  5642	   extensionHost
    0	    33	  7515	     /Applications/Visual Studio Code.app/Contents/MacOS/Electron --ms-enable-electron-run-as-node /Applications/Visual Studio Code.app/Contents/Resources/app/extensions/json-language-features/server/dist/node/jsonServerMain --node-ipc --clientProcessId=5642
    0	   328	 53115	     /Users/michaeldeeks/go/bin/gopls -mode=stdio
    5	   492	 39854	   window (rg --files --hidden --case-sensitive -g .sh — carrot [SSH: bento])
    0	   197	 39856	   extensionHost
    0	    66	 39880	     /Applications/Visual Studio Code.app/Contents/MacOS/Electron --ms-enable-electron-run-as-node /Users/michaeldeeks/.vscode/extensions/ms-vscode-remote.remote-ssh-0.76.1/out/localServer.js {"serverId":1,"ipcHandlePath":"/var/folders/kl/ftwb65qj15s117sd11820xk40000gp/T/vscode-ssh-askpass-0fa9ad4c728961d969e6309de1b1cbbe9345831c.sock","sshCommand":"ssh","sshArgs":["-v","-T","-D","54315","bento"],"serverDataFolderName":".vscode-server","dataFilePath":"/Users/michaeldeeks/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-83a2ce0f-c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1-0.76.1/data.json"}
    0	     0	 39895	       ssh -v -T -D 54315 bento
   11	   295	 41497	   window (carrot)
   16	   655	 41499	   extensionHost
    0	    33	 41517	     /Users/michaeldeeks/go/bin/gopls -mode=stdio
   29	    66	 41792	   issue-reporter
    0	   328	 73740	   window (code — bento-remote)
    0	   197	 73741	   extensionHost

Remote: SSH: bento
```

</details>
<details>
<summary>Workspace Info</summary>

```
|  Window (carrot)
|  Window (server.go — bento)
|  Window (code — bento-remote)
|  Window (main.go — bento-server)
|    Folder (carrot): more than 20823 files
|      File types: rb(2666) py(2183) sql(2163) ts(1940) tsx(1518) go(1491)
|                  yml(608) yaml(592) csv(403) sh(222)
|      Conf files: dockerfile(181) makefile(40) github-actions(30)
|                  package.json(7) tsconfig.json(4) webpack.config.js(2)
|    Folder (bento): 1274 files
|      File types: go(779) yml(55) sh(41) tmpl(22) md(21) json(20) ts(17)
|                  tsx(12) png(8) gitignore(6)
|      Conf files: makefile(5) github-actions(4) tsconfig.json(3)
|                  dockerfile(2) package.json(2)
|    Folder (bento-remote): more than 20963 files
|      File types: py(4231) pyc(4229) yaml(2611) yml(1505) rst(624) cfg(232)
|                  txt(191) json(187) j2(125) sh(102)
|      Conf files: settings.json(2) makefile(2)
|    Folder (bento-server): 138 files
|      File types: go(68) sh(8) yml(5) mod(3) sum(3) gitignore(2) md(2)
|                  bash(2) editorconfig(1) development(1)
|      Conf files: dockerfile(4) makefile(1);
```

</details>
<details><summary>Extensions (23)</summary>

Extension|Author (truncated)|Version
---|---|---
terraform|4op|0.2.2
gitlens|eam|12.0.5
codespaces|Git|1.5.7
go|gol|0.32.0
terraform|has|2.20.1
vscode-icon-theme|jtl|1.6.6
keen-neutral-icon-theme|kee|1.0.8
hjson|lak|0.3.2
vscode-docker|ms-|1.20.0
python|ms-|2022.2.1924087327
vscode-pylance|ms-|2022.3.2
jupyter|ms-|2022.2.1030672458
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.6
remote-containers|ms-|0.224.3
remote-ssh|ms-|0.76.1
remote-ssh-edit|ms-|0.76.1
remote-wsl|ms-|0.64.2
vscode-remote-extensionpack|ms-|0.21.0
ruby|reb|0.28.1
shellcheck|tim|0.18.9
vscode-icons-mac|way|7.25.3
vscode-ruby|win|0.28.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30256859
pythonvspyl392:30443607
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vsclayoutctrt:30451275
dsvsc009:30452663
pythonvsnew555:30450902
vscscmwlcmt:30438805
vscgsvidc:30447482
dockersurvey3cf:30453610
pynewfile477cf:30450039

```

</details>

**Comment by roblourens** [comment]
You will see a rg process every time because we start one in the background to refresh the cache on each search. But the 2nd+ search itself will not be blocked on that rg process. I just double-checked in a big workspace that this is still the case. 
If you watch the search and the rg process, you should see the search complete before the rg process at least. You think that searches after the first one take pretty much the same amount of time though?

**Comment by mdeeks** [comment]
Ahh I see. That makes sense thanks. I did some testing and recorded it. The results are inconsistent. Sometimes searches are instant (no loading at all) and sometimes they pause for loading anywhere from 1-4 seconds.

For some extra context: 
- Our repo uses symlinks and unfortunately we can't turn off following symlinks in vscode. 
- The result of the rg search above outputs 673k files (65MB content)
- The example file I'm searching for below "bento.yml" should produce only about 45 results

Test from the video:
1. On a fresh window (after waiting a minute for it to load) I searched for "bento". Since thats the first blocking search it takes about 5 seconds.
2. Subsequent searches for "bento" are nearly instant 👍 
3. Searching for "bento." takes about 4 seconds
4. Typing out the rest of the filename to "bento.yml" is instant
5. Deleting the search and searching again for "bent" takes about 4 seconds

I do a couple of more searches for various spellings of "config" and some are instant while some are slow ("config" is fast, "conf" is slow)

https://user-images.githubusercontent.com/2966467/159309557-0857018e-6cb2-4e5f-a212-c71992f7ef34.mov

**Comment by roblourens** [comment]
Some of the fast/slow might just be from the number of results. It's been awhile since I dug into this, but I think our cache is smart enough cache by filter prefix, so typing one more character would be fast. I don't think there is something special about adding the `.`, but I can double-check. In my testing in a not-quite-as-large repo, I don't see it getting slower when I add a `.`.

In my testing, the size of the result set makes a big difference in the response time. There's a max of 512 results.

**Comment by captainbeardo** [comment]
I am having similar performance issues. it's about 5s before I get a result. Workspace is approximately 1000 files. Even subsequent searches for the same terms are not any faster. I'm using remote-ssh instances of VS code.

**Comment by roblourens** [comment]
@captainbeardo you could see whether https://github.com/microsoft/vscode/wiki/Search-Issues helps

**Comment by captainbeardo** [comment]
@roblourens turning off following sym links helped a little, which is strange because there are no symlinks in the directory. I turned on logging and captured the ripgrep command. Running that separately returns near immediate results: 

```
$ <redacted>/.vscode-server/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node_modules/@vscode/ripgrep/bin/rg --files --hidden --case-sensitive -g '!**/.git' -g '!**/.svn' -g '!**/.hg' -g '!**/CVS' -g '!**/.DS_Store' -g '!**/Thumbs.db' -g '!**/node_modules' -g '!**/bower_components' -g '!**/*.code-search' --no-ignore-parent --no-config --no-ignore-global | wc -l
1093

real    0m0.011s
user    0m0.000s
sys     0m0.035s
```

On my local machine (Mac) a Code Helper (Render) process spikes CPU usage while it's doing the search. There's latency when typing as well. Nothing on the remote host shows significant CPU usage.

**Comment by roblourens** [comment]
Can you share the log? Which process is showing high cpu (in Help > Open Process Explorer)? If it's the window process, could you grab a trace from Help > Toggle Developer Tools, then use the Performance tab to grab a trace?

**Comment by roblourens** [comment]
It will have file paths, but that's it. You can zip it and upload it here, or email it to me.

**Comment by captainbeardo** [comment]
Here it is: [slow-quick-open.json.zip](https://github.com/microsoft/vscode/files/9283920/slow-quick-open.json.zip)

**Comment by roblourens** [comment]
Thanks. Looks like we are spending lots of time in `findFormatting`, and it actually looks like we were discussing the same issue in https://github.com/microsoft/vscode/issues/157056 and it should be fixed in Insiders already.

**Comment by captainbeardo** [comment]
@roblourens Can confirm, it is performant on the latest Insiders build. Thanks!
